File: AVTProject1/AVTProject1/src/Entity.cpp
#include "Entity.h"

namespace AVTEngine
{
	Entity::Entity() {}

	Entity::Entity(SceneNode *node_, Shader *shader_, Mesh *mesh_):
		node(node_), shader(shader_), mesh(mesh_){
	}

	Entity::Entity(Shader *shader_, Mesh *mesh_, float maxVelocity_, float maxTurnRate_, glm::vec3 orientation_){
		enabled = true;
		rotationAccum = 0;
		orientation = orientation_;
	};

	//TODO teste
	Entity::Entity(SceneNode *node_) : 
		node(node_) {
		enabled = true;
		rotationAccum = 0;
	};

	void Entity::disable() {
		enabled = false;
	};

	void Entity::enable() {
		enabled = true;
	};

	//Uma entidade disabled nao "existe" no mapa, ou seja, nao sofre colisoes e nao e atualizada (apesar de ainda poder aparecer)
	bool Entity::isDisabled() {
		return !enabled;
	}

	glm::vec3 Entity::getOrientation() {
		return orientation;
	}

	glm::vec3 Entity::getPosition() {
		return position;
	}

	void Entity::setPosition(glm::vec3 position_) {
		position = position_;
		node->setPosition(position_); //Tell node to change position
	}


	//Aplicar rotaçao na matriz do modelo
	/* 
	void Entity::rotate(float angle_) {
		rotationAccum += angle_;
		glm::vec3 y_axis = Y_AXIS;

		//Rotate model
		ModelViewMatrix = glm::rotate(ModelViewMatrix, angle_, y_axis);
		orientation = glm::rotateY(orientation, angle_);
	}
	*/
	void Entity::rotate(float angle_) {

		rotationAccum += angle_;
		glm::vec3 y_axis = Y_AXIS;

		//Rotate model
		glm::quat quaternion = glm::angleAxis(rotationAccum, y_axis);
		node->setRotation(quaternion);
		
		//Also update orientation, for movement calculation purposes
		orientation = glm::rotateY(orientation, angle_);
	}

	void Entity::orangeRotate(float angle_, glm::vec3 axis_) {

		rotationAccum += angle_;

		//Rotate model
		glm::quat quaternion = glm::angleAxis(rotationAccum, axis_);
		node->setRotation(quaternion);

		//Also update orientation, for movement calculation purposes
		orientation = glm::rotateY(orientation, angle_);
	}

	void Entity::setMesh(Mesh* mesh_) {
		mesh = mesh_;
	}

	void Entity::setMaterial(Material* material_) {
		material = material_;
	}

	void Entity::update(float delta_)
	{
		//Empty
	}//Virtual

}

File: WebGL/js/entity.js
"use strict"

// Superclasse de todos os objectos que se movem
class Entity{

	constructor(node){
		this.object = new THREE.Object3D();
		this.rotationAccum = 0;
		this.node = node;
		this.enabled = true;

		this.position = new THREE.Vector3(0,0,0);
		this.orientation = new THREE.Vector3(0,0,0);
		this.movementOrientation = new THREE.Vector3(0,0,0);
		this.rotation = new THREE.Quaternion(0, 0, 0, 0);
	}

	disable(){
        this.enabled = false;
    }

    enable(){
        this.enabled = true;
    }

    // uma entidade disabled nao "existe" no mapa, ou seja, nao sofre colisoes e nao e atualizada (apesar de ainda poder aparecer)
    isDisabled(){
    	return !this.enabled;
    }

	getOrientation(){
		return this.orientation.clone();
	}

	getPosition(){
		//return this.position.clone();
		return this.object.position.clone();
	}

	setPosition(position){
		this.object.position.copy(position);
		this.position.copy(position);
		this.node.position = this.position;
		this.node.visible = this.visible;
		//this.node.dirty = true;
	}

	rotate(angle){
		this.rotationAccum += angle;

		//Rotate model
		this.object.rotateOnAxis(Y_AXIS, angle);
		this.object.getWorldQuaternion(this.rotation);
		this.node.rotation = this.rotation;
		//this.node.dirty = true;

		//Also update orientation, for movement calculation purposes
		this.orientation.applyAxisAngle(Y_AXIS, angle);	
	}

	orangeRotate(angle, axis) {

		this.rotationAccum += angle;

		//Rotate model
		this.object.rotateOnAxis(axis, angle);
		this.object.getWorldQuaternion(this.rotation);
		this.node.rotation = this.rotation;

		//Also update orientation, for movement calculation purposes
		this.orientation.applyAxisAngle(axis, angle);	
	}

}